### PR TITLE
Better errors for partially applied combinators

### PR DESCRIPTION
**Pull request by acondolu**
#### Motivation
Expressions with partially applied combinators are actually well-typed, even though they do not represent valid endpoints.

Take for instance the following malformed API:
```haskell
> type BadAPI = CaptureAll "path" :> EmptyAPI
```
which is incorrect because `CaptureAll` requires an additional argument (the type of the values to capture). This definition does not produce a type error, but later when trying to use `BadAPI` one gets confusing errors due to non-existing instances:
```haskell

> err :: (Link -> a) -> Proxy BadAPI -> Link -> MkLink BadAPI a
> err = toLink
```
```
  * No instance for (HasLink BadAPI) arising from a use of `toLink'
  * In the expression: toLink
  ...
```
The goal is to have more explicative error messages.

#### Description of the solution
This PR proposes custom error messages at the type level. Basically, it adds one new instance for each non-existing instance due to combinators that are not fully-applied. These new instances have a type error as constraint, so to show a more precise message to the user.

For instance, the error shown above is caught by the following erroring instance:
```haskell
instance TypeError (...)
    => HasLink (CaptureAll sym :> sub)
  where ...
```
Compare the error message shown above with the following, resulting one:
```
  * There is no instance for HasLink (CaptureAll "path" :> ...)
    CaptureAll expects 2 arguments
  * In the expression: toLink
  ...
```

Currently, the code in the PR defines instances only for `HasLink` w.r.t. `CaptureAll`. If the mantainers find this approach valid, similar instances will be defined for other combinators and classes/families (e.g. `IsElem`).

#### Alternatives
The approach suggested above provides much nicer error messages; the only downside is code duplication (it requires a new instance for each class, combinator, and `0...arity-1`). The abbreviations defined in `TypeErrors.hs` mitigate the problem, and allow reuse.

An alternative would be to have a single catch-all instance for all partially-applied combinators, like the following (inspired from the one [here](https://github.com/haskell-servant/servant/blob/ec0cd8a9471da321cac127a36adb1f7c138b658c/servant-server/src/Servant/Server/Internal.hs#L808)):
```haskell
instance TypeError (...)
    => HasLink ((arr :: k -> l) :> sub)
  where ...
```
which only checks for generic unapplied stuff, and results in a less informative message:
```
  * There is no instance for HasLink (CaptureAll "path" :> ...)
  * In the expression: toLink
  ...
```

### CODE REVIEW

**Comment by acondolu** [comment]
Hello,

Any updates on this PR? Please let me know if you have any comment.

File: servant/src/Servant/Links.hs [hunk 1]
```diff
@@ -579,3 +579,46 @@ simpleToLink _ toA _ = toLink toA (Proxy :: Proxy sub)
...
+    type MkLink (CaptureAll :> sub) a = TypeError (HasLinkError CaptureAll CaptureAll)
+    toLink = error "unreachable"
+
+instance TypeError (HasLinkError CaptureAll (CaptureAll sym))
+    => HasLink (CaptureAll sym :> sub)
+  where
+    type MkLink (CaptureAll sym :> sub) a = TypeError (HasLinkError CaptureAll (CaptureAll sym))
+    toLink = error "unreachable"
+
+-- TODO: Errors for other combinators.
```
[the remark this answers is not in the document]
Comment (acondolu) [reply servant/src/Servant/Links.hs:638]: Sure: #1461 🎉

**Comment by acondolu** [comment]
A colleague of mine pointed out that, to be able to use the catch-all instance with an `OVERLAPPABLE` pragma, it suffices to avoid supplying the associate type instance. Then GHC is happy. For instance:

```haskell
instance {-# OVERLAPPABLE #-} TypeError (NoInstanceFor (HasLink (arr :> "..."))) => HasLink (arr :> sub)

instance {-# OVERLAPPABLE #-} TypeError (NoInstanceFor' (HasLink api)) => HasLink api
```

**Comment by gdeest** [comment]
OK, I finally had a good look and am finding this quite reasonable. I don't see any reasonable way to reduce verbosity any further.

If no one has any objection, I am going to use  this approach to other combinators and merge.

**Comment by tchoutri** [comment]
+1

**Comment by gdeest** [comment]
@acondolu It looks like GHC 8.6 is not happy with the catchall instance :(

```haskell
 Error:     • Type indexes must match class instance head
      Expected: MkLink (arr :> sub) <tv>
        Actual: MkLink (arr :> _1) _2
      where the `<tv>' arguments are type variables,
      distinct from each other and from the instance variables
    • In the type instance declaration for ‘MkLink’
      In the instance declaration for ‘HasLink ((arr :: a -> b) :> sub)’
```

**Review by gdeest — APPROVED** [review]
Let's merge this, and extend the approach in follow-up PRs.

Thanks @acondolu for your contribution ! I believe it is a very nice usability enhancement.

**Pull request merged**